File: FBControlCore/Applications/FBBundleDescriptor+Application.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBBundleDescriptor+Application.h"

#import "FBArchiveOperations.h"
#import "FBBinaryDescriptor.h"
#import "FBCollectionInformation.h"
#import "FBControlCoreError.h"
#import "FBControlCoreError.h"
#import "FBControlCoreGlobalConfiguration.h"
#import "FBControlCoreLogger.h"
#import "FBProcess.h"
#import "FBProcessBuilder.h"

@implementation FBBundleDescriptor (Application)

#pragma mark Public

+ (FBBundleDescriptor *)findAppPathFromDirectory:(NSURL *)directory logger:(id<FBControlCoreLogger>)logger error:(NSError **)error
{
  NSDirectoryEnumerator *directoryEnumerator = [NSFileManager.defaultManager
    enumeratorAtURL:directory
    includingPropertiesForKeys:@[NSURLIsDirectoryKey]
    options:0
    errorHandler:nil];
  NSMutableArray<NSString *> *applicationPaths = NSMutableArray.array;
  NSMutableArray<NSString *> *nonApplicationPaths = NSMutableArray.array;
  [logger logFormat:@"Finding Application Path from root directory %@", directory];
  for (NSURL *fileURL in directoryEnumerator) {
    NSString *path = fileURL.path;
    if ([FBBundleDescriptor isApplicationAtPath:path]) {
      [logger logFormat:@"Found application at path %@", path];
      [applicationPaths addObject:path];
      [directoryEnumerator skipDescendants];
    } else {
      [logger logFormat:@"Non-application path at %@", path];
      [nonApplicationPaths addObject:path];
    }
  }
  if (applicationPaths.count == 0) {
    return [[FBControlCoreError
      describeFormat:@"Could not find an Application in IPA, present files %@", [FBCollectionInformation oneLineDescriptionFromArray:[nonApplicationPaths valueForKey:@"lastPathComponent"]]]
      fail:error];
  }
  if (applicationPaths.count > 1) {
    return [[FBControlCoreError
      describeFormat:@"Expected only one Application in IPA, found %lu: %@", applicationPaths.count, [FBCollectionInformation oneLineDescriptionFromArray:[applicationPaths valueForKey:@"lastPathComponent"]]]
      fail:error];
  }
  NSString *applicationPath = applicationPaths.firstObject;
  [logger logFormat:@"Using Application at path %@", applicationPath];
  FBBundleDescriptor *bundle = [FBBundleDescriptor bundleFromPath:applicationPath error:error];
  [logger logFormat:@"Bundle in IPA is %@", bundle];
  if (!bundle) {
    return nil;
  }
  return bundle;
}

+ (BOOL)isApplicationAtPath:(NSString *)path
{
  BOOL isDirectory = NO;
  return path != nil
    && [path hasSuffix:@".app"]
    && [[NSFileManager defaultManager] fileExistsAtPath:path isDirectory:&isDirectory]
    && isDirectory;
}

@end

File: FBControlCore/Applications/FBBundleDescriptor+Application.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

extension FBBundleDescriptor {

  @objc(findAppPathFromDirectory:logger:error:)
  public class func findAppPath(fromDirectory directory: URL, logger: FBControlCoreLogger?) throws -> FBBundleDescriptor {
    let directoryEnumerator = FileManager.default.enumerator(
      at: directory,
      includingPropertiesForKeys: [.isDirectoryKey],
      options: [],
      errorHandler: nil
    )
    var applicationPaths: [String] = []
    var nonApplicationPaths: [String] = []
    logger?.log("Finding Application Path from root directory \(directory)")
    if let enumerator = directoryEnumerator {
      for case let fileURL as URL in enumerator {
        let path = fileURL.path
        if FBBundleDescriptor.isApplication(atPath: path) {
          logger?.log("Found application at path \(path)")
          applicationPaths.append(path)
          enumerator.skipDescendants()
        } else {
          logger?.log("Non-application path at \(path)")
          nonApplicationPaths.append(path)
        }
      }
    }
    if applicationPaths.isEmpty {
      let lastComponents = nonApplicationPaths.map { ($0 as NSString).lastPathComponent }
      throw FBControlCoreError
        .describe("Could not find an Application in IPA, present files \(FBCollectionInformation.oneLineDescription(from: lastComponents))")
        .build()
    }
    if applicationPaths.count > 1 {
      let lastComponents = applicationPaths.map { ($0 as NSString).lastPathComponent }
      throw FBControlCoreError
        .describe("Expected only one Application in IPA, found \(applicationPaths.count): \(FBCollectionInformation.oneLineDescription(from: lastComponents))")
        .build()
    }
    let applicationPath = applicationPaths[0]
    logger?.log("Using Application at path \(applicationPath)")
    let bundle = try FBBundleDescriptor.bundle(fromPath: applicationPath)
    logger?.log("Bundle in IPA is \(bundle)")
    return bundle
  }

  @objc(isApplicationAtPath:)
  public class func isApplication(atPath path: String) -> Bool {
    var isDirectory: ObjCBool = false
    return path.hasSuffix(".app")
      && FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory)
      && isDirectory.boolValue
  }
}
